xcdat/regridder/grid.py

```diff
@@ -425,94 +426,106 @@ def create_zonal_grid(grid: xr.Dataset) -> xr.Dataset:
     # Ignore `Argument 1 to "create_grid" has incompatible type
     # "Union[Dataset, DataArray]"; expected "Union[ndarray[Any, Any], DataArray]"
     # mypy(error)` because this arg is validated to be a DataArray beforehand.
-    return create_grid(lat, out_lon_data, lat_bnds=lat_bnds, lon_bnds=lon_bnds)  # type: ignore
+    return create_grid(lat=(lat, lat_bnds), lon=(out_lon_data, lon_bnds))  # type: ignore
 
 
-def create_grid(
-    lat: Union[np.ndarray, xr.DataArray],
-    lon: Union[np.ndarray, xr.DataArray],
-    lat_bnds: Optional[Union[np.ndarray, xr.DataArray]] = None,
-    lon_bnds: Optional[Union[np.ndarray, xr.DataArray]] = None,
-) -> xr.Dataset:
-    """Creates a grid for a give latitude and longitude array.
+def create_grid(**kwargs: CoordOptionalBnds) -> xr.Dataset:
+    """Creates a grid from coordinate mapping.
 
     Parameters
     ----------
-    lat : Union[np.ndarray, xr.DataArray]
-        Array of latitude values.
-    lon : Union[np.ndarray, xr.DataArray]
-        Array of longitude values.
-    lat_bnds : Optional[Union[np.ndarray, xr.DataArray]]
-        Array of bounds for latitude values.
-    lon_bnds : Optional[Union[np.ndarray, xr.DataArray]]
-        Array of bounds for longitude values.
+    **kwargs : CoordOptionalBnds
+        Mapping of coordinate name and data with optional bounds. See
+        :py:data:`xcdat.axis.VAR_NAME_MAP` for valid coordinate names.
```

Comment:
Hey Jason, my suggestion here was to make the parameters for `create_grid` more explicit. 

In my opinion, using `**kwargs` over an explicit args/kwargs can create confusion and reduce readability since there isn't set number of clearly defined parameters for the function. Here is a post a read about `**kwargs`: https://softwareengineering.stackexchange.com/a/294486

I would suggest keeping the args and kwargs, then extending as needed.

Reply:
@pochedls @chengzhuzhang @lee1043 @tomvothecoder 
How do we feel about the following function to replace the current `create_grid`.
```python
def create_grid(x: GridCoord, y: GridCoord, z: GridCoord) -> xr.Dataset:
```
Where `GridCoord` could be any of the following forms.
```
(name, data)
(name, data, attrs)
(name, data, bnds)
(name, data, bnds, attrs)
```

Reply:
From what I understand, this is a tuple where:
- name = string
- data = array of floats
- attrs = dictionary mapping str to str <OPTIONAL>
- bnds = 2d array of floats <OPTIONAL>

Is this correct?

Reply:
@tomvothecoder Yes that's correct. The `data` and `bnds` values can be a list, ndarray or DataArray (1d/2d respectively).

Reply:
Thanks for confirming. I am curious about the benefits you see in using a tuple of sub-arguments over an `xr.DataArray`? I'm pretty sure this was discussed with the team during our meetings, but I didn't really weigh in because there's a lot to track!

Reply:
The only benefit I can see is for users who may not want learn how to construct a `xr.DataArray`. This was discussed, I remember it being opposed at first but thought we'd changed the end decision to allow more than just `xr.DataArray`. I'm fine just make the arguments `xr.DataArray` and updating the docstring to include creating the `xr.DataArray`'s.

Reply:
I agree about only seeing that benefit. The tradeoff of API flexibility with simplicity and maintainability doesn't seem enough in my opinion. I'm in favor with keeping the arguments as `xr.DataArray`. This approach encourages the use of xarray objects (this library is an extension of xarray) and keeps the implementation and API usage simple. I noticed other libraries like [xgcm](https://xgcm.readthedocs.io/en/latest/api.html) and [xclim](https://xclim.readthedocs.io/en/stable/index.html) focus on accepting xarray objects.

Reply:
What if we include a `create_axis()` function (similar to the one at bottom)?

Then we could pass in an axis (DataArray derived from `create_axis()`) or a tuple of `(axis, bnds)`. The bnds could be an array or dataarray (if dataarray, we add `bnds.name` as the `bounds` attribute to `axis`, otherwise the bounds name would be `axis.name + "_bnds"` and would be produced automatically). 

```python
def create_axis(aid, axis_values, attrs=None):
    # define default attributes
    lat_attrs = {'units': 'degrees_north',
                 'axis': 'Y',
                 'long_name': 'latitude',
                 'standard_name': 'latitude'}
    lon_attrs = {'units': 'degrees_east',
                 'axis': 'X',
                 'long_name': 'longitude',
                 'standard_name': 'longitude'}
    plev_attrs = {'units': 'Pa',
                  'axis': 'Z',
                  'long_name': 'pressure',
                  'standard_name': 'air_pressure',
                  'positive': 'down'}
    lev_attrs = {'units': 'm',
                  'axis': 'Z',
                  'long_name': 'ocean depth coordinate',
                  'standard_name': 'depth',
                  'positive': 'down'}
    default_attrs = {'lat': lat_attrs,
                     'latitude': lat_attrs,
                     'lon': lon_attrs,
                     'longitude': lon_attrs,
                     'lev': lev_attrs,
                     'plev': plev_attrs}
    if attrs is None:
        attrs = default_attrs[aid]

    newAxis = xr.DataArray(
        name=aid,
        data=axis_values,
        dims=[aid],
        coords={aid: axis_values},
        attrs=attrs)

    return newAxis
```

Reply:
Are there use cases where `create_axis` would be used without `create_grid`?

Reply:
Yeah – I have instances where I have a numpy array and want to create a DataArray (with correct axis information). I think this is kind of common. @mzelinka – were you saying that you have cases where it would be helpful to generate a new axis (with correct metadata/attributes)?

Reply:
> What if we include a create_axis() function (similar to the one at bottom)?

I can see a function like `create_axis` making the process of creating xr.DataArray axes faster and less prone to error. 
**Maybe this can be scoped out in another GitHub issue?**

I had a few questions (for myself too):

  - When and how often might users want/need to create axes instead of using the ones in the file?
  - How should we decide what is considered the most sensible default attributes for each axis? Xarray sets some like `calendar="standard"` for the time axis, and we set a few others like units for latitude/longitude.
  - If we decide this makes sense to implement, we should consider calling it `create_dim` since we have [`get_dim_keys()`](https://xcdat.readthedocs.io/en/stable/generated/xcdat.get_dim_keys.html#xcdat.get_dim_keys) and [`get_dim_coords()`](https://xcdat.readthedocs.io/en/stable/generated/xcdat.get_dim_coords.html#xcdat.get_dim_coords).



---

I also like the idea of having a static dictionary to store default CF attributes. We have default attrs set in various parts of the code-base (e.g., calendar=`standard` for the time axis). It'll be cleaner to have a single reference point for default attrs.

```python
# axis.py
CF_DEFAULT_ATTRS = {
	"lat": '{'units': 'degrees_north',
                 'axis': 'Y',
                 'long_name': 'latitude',
                 'standard_name': 'latitude'
     },
	....
}
```

Reply:
@pochedls @tomvothecoder Lets move this discussion to the [new issue](https://github.com/xCDAT/xcdat/issues/496).